### PR TITLE
Add getopt support to src/nstack.c

### PR DESCRIPTION
**Pull request by kevinbird61**
Start to trace the code, I think it will be nice if we can using getopt to configure our parameters.

### CODE REVIEW

**Comment by jserv** [comment]
If you are willing to add IPv6 support, please send another dedicated pull request.

File: src/nstack.c [hunk 1]
```diff
@@ -326,30 +326,59 @@ void nstack_stop(void)
 
 int main(int argc, char *argv[])
 {
+    /* getopt support */
+    int opt; 
+    char *intf, *mac;
+    while((opt = getopt(argc, argv, "i:m:")) != -1){
+        switch (opt) {
+            case 'i':
+                // interface, ex: veth0, eth0 ...
```
Comment (jserv) [inline src/nstack.c]: Use C style comment. That is, `/* ... */`.

File: src/nstack.c [hunk 2]
```diff
@@ -326,30 +326,59 @@ void nstack_stop(void)
...
 int main(int argc, char *argv[])
 {
+    /* getopt support */
+    int opt; 
+    char *intf, *mac;
+    while((opt = getopt(argc, argv, "i:m:")) != -1){
+        switch (opt) {
+            case 'i':
+                // interface, ex: veth0, eth0 ... 
+                // intf = malloc(sizeof(char)*strlen(optarg)+1);
```
Comment (jserv) [inline src/nstack.c]: Don't leave the dead code.

File: src/nstack.c [hunk 3]
```diff
@@ -326,30 +326,59 @@ void nstack_stop(void)
...
+                // mac Address
+                mac = optarg;
+                break;
+            default: /* not match */
+                fprintf(stderr, "Usage: %s [-i <interface>] [-m <mac_addr>] [-h]\n",argv[0]);
+                exit(EXIT_FAILURE);
+        }
+    }
+    // argv[1] = Network Interface, ex: eth0, veth1, ... etc.
+    /* orig
```
Comment (jserv) [inline src/nstack.c]: Ditto. Dead code.

File: src/nstack.c [hunk 4]
```diff
@@ -326,30 +326,59 @@ void nstack_stop(void)
...
     sigprocmask(SIG_SETMASK, &sigset, NULL);
 
+
     handle = ether_init(ether_args);
     if (handle == -1) {
         perror("Failed to init");
         exit(1);
     }
 
+    // 167772162    => 10.0.0.2
```
Comment (jserv) [inline src/nstack.c]: You shall send another pull request to append or improve comments.

File: tools/run.sh [hunk 5]
```diff
@@ -4,4 +4,4 @@
 dd if=/dev/zero of=/tmp/unetcat.sock bs=1024 count=1024
 
 sudo setcap cap_net_raw,cap_net_admin,cap_net_bind_service+eip build/inetd
-sudo ip netns exec TEST su $USER -c "build/inetd $1"
+sudo ip netns exec TEST su $USER -c "build/inetd -i $1"
```
Comment (jserv) [inline tools/run.sh]: What is the intention?

File: src/linux/ether.c [hunk 6]
```diff
@@ -132,21 +132,47 @@ int ether_init(char *const args[])
...
+        // errno = ENOTSUP;
+        // return -1;
+        if(ioctl(eth->el_fd, SIOCGIFHWADDR, &args[1]) < 0){
+            goto fail;
+        }
+
+        int mac_seg[6];
+        // adding "%c" on the end of format string, to secure there are no
+        // extraneous char after the sixth octet
+        if(6 == sscanf(args[1], "%x:%x:%x:%x:%x:%x%*c",
```
Comment (jserv) [inline src/linux/ether.c]: Use `clang-format -i` to indent the source files.

File: src/nstack.c [hunk 7]
```diff
@@ -326,30 +326,59 @@ void nstack_stop(void)
...
+            case 'm':
+                // mac Address
+                mac = optarg;
+                break;
+            default: /* not match */
+                fprintf(stderr, "Usage: %s [-i <interface>] [-m <mac_addr>] [-h]\n",argv[0]);
+                exit(EXIT_FAILURE);
+        }
+    }
+    // argv[1] = Network Interface, ex: eth0, veth1, ... etc.
```
Comment (jserv) [inline src/nstack.c]: You shall not use "ex" here. Instead, the abbreviation e.g. stands for Latin exempli gratia, which means "for example. "

File: tools/run.sh (hunk 5 above)
Comment (jserv) [reply tools/run.sh]: What if no interface is found?

Comment (kevinbird61) [reply tools/run.sh]: @jserv  I've used the original structure - `ether_args`  array in [src/nstack.c, line 348](https://github.com/kevinbird61/nstack/blob/master/src/nstack.c#L348), only replace `argv[1]` with `intf`(**Network Interface**) and `NULL` with `mac`(**MAC address**).

And if no specify any interface for `inetd`, it will display the error message:
`Failed to init: Operation not permitted`

**Comment by jserv** [comment]
Use `git rebase -i` to rework these commits.

**Comment by kevinbird61** [comment]
Ok, I've used `git rebase -i` to refactor the latest 3 commits.
Thank for reviewing! And I will continue to work on `ipv6`.

**Comment by jserv** [comment]
The commit `Add ipv6 header structure` should be decoupled from this pull request. The merge sequence should be:
1. improve comments;
2. add IPv6 header;
3. add `getopt` to reflect IPv6 changes;

**Comment by jserv** [comment]
@kevinbird61, send 3 pull request independently.

**Comment by jserv** [comment]
@laochanlam, are you willing to rework this pull request?

**Comment by laochanlam** [comment]
@jserv I would like to improve the performance of TCP protocol on nstack first.

**Comment by jserv** [comment]
@laochanlam, It sounds great. You can create an new issue to describe what you are working and the area you already pay attention to.
